### PR TITLE
Add new starter code for all languages

### PR DESCRIPTION
**Pull request by glokta1**
To be compatible with  program interface overhaul changes.

### CODE REVIEW

File: compiled_starters/c/.codecrafters/compile.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+#
+# This script is used to compile your program on CodeCrafters
+# 
+# This runs before .codecrafters/run.sh
+#
+# Learn more: <insert-link>
```
Comment (rohitpaulk) [inline compiled_starters/c/.codecrafters/compile.sh:7]: Link not filled

File: compiled_starters/cpp/.codecrafters/compile.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+#
+# This script is used to compile your program on CodeCrafters
+# 
+# This runs before .codecrafters/run.sh
+#
+# Learn more: <insert-link>
+
+cmake -B build -S . -DCMAKE_TOOLCHAIN_FILE=${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake
```
Comment (rohitpaulk) [inline compiled_starters/cpp/.codecrafters/compile.sh:9]: This needs `set -e` too (maybe we could just add this to all scripts?)

File: compiled_starters/c/your_program.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+#
+# Use this script to run your program LOCALLY.
+#
+# Note: Changing this script WILL NOT affect how CodeCrafters runs your program.
+#
+#
```
Comment (rohitpaulk) [inline compiled_starters/c/your_program.sh:7]: We don't need the double empty lines here

File: compiled_starters/cpp/your_program.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+#
+# Note: Changing this script WILL NOT affect how CodeCrafters runs your program.
+#
+#
+# Learn more: docs.codecrafters.io/program-interface
+
+# Exit early if any commands fail
+set -e
+
+CODECRAFTERS_SUBMISSION_DIR=$(dirname 0)
```
Comment (rohitpaulk) [inline compiled_starters/cpp/your_program.sh:13]: This is present for C++ but not C - let's pick one and be consistent

File: compiled_starters/go/.codecrafters/compile.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+#
+# This script is used to compile your program on CodeCrafters
+# 
+# This runs before .codecrafters/run.sh
+#
+# Learn more: <insert-link>
+
+go build -o /tmp/shell-target cmd/myshell/*.go
```
Comment (rohitpaulk) [inline compiled_starters/go/.codecrafters/compile.sh:9]: All `compile.sh` scripts need `set -e` at the beginning. Else a failure won't cause a non-zero exit code, and we won't know when the compile step fails.

File: compiled_starters/java/src/main/java/Main.java [hunk 6]
```diff
@@ -1,12 +1,11 @@
...
 
 public class Main {
     public static void main(String[] args) throws Exception {
         // Uncomment this block to pass the first stage
         // System.out.print("$ ");
-        //
-        // Scanner scanner = new Scanner(System.in);
-        // String input = scanner.nextLine();
+
+        Scanner scanner = new Scanner(System.in);
```
Comment (rohitpaulk) [inline compiled_starters/java/src/main/java/Main.java:8]: @glokta1 is this change intentional? Don't think this was uncommented earlier

File: compiled_starters/javascript/.codecrafters/compile.sh [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+#
+# This script is used to compile your program on CodeCrafters
+# 
+# This runs before .codecrafters/run.sh
+#
+# Learn more: <insert-link>
+
```
Comment (rohitpaulk) [inline compiled_starters/javascript/.codecrafters/compile.sh:8]: If there is no compile step, we should a comment. Suggested text: 

```
# ... 
# 
# Learn more: ...

# (This file is empty since JavaScript programs don't use a compile step)
```

File: compiled_starters/go/.codecrafters/run.sh [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+#
+# This script is used to run your program on CodeCrafters
+# 
+# This runs after .codecrafters/compile.sh
```
Comment (rohitpaulk) [inline compiled_starters/go/.codecrafters/run.sh:5]: Some files say "(if required)" on this line, others don't - we need to be consistent

File: compiled_starters/rust/src/main.rs [hunk 9]
```diff
@@ -6,7 +6,6 @@ fn main() {
     // print!("$ ");
     // io::stdout().flush().unwrap();
 
-    // Wait for user input
```
Comment (rohitpaulk) [inline compiled_starters/rust/src/main.rs:9]: @glokta1 why are there unrelated changes like this here?

File: dockerfiles/java-21.Dockerfile [hunk 10]
```diff
@@ -15,5 +15,5 @@ RUN mkdir -p /app-cached
 RUN mv /app/target /app-cached # Is this needed?
 
 # Pre-compile steps
-RUN printf "cd \${CODECRAFTERS_SUBMISSION_DIR} && mvn -B package -Ddir=/tmp/codecrafters-shell-target && sed -i 's/^\(mvn .*\)/#\1/' ./your_shell.sh" > /codecrafters-precompile.sh
+RUN printf "cd \${CODECRAFTERS_SUBMISSION_DIR} && mvn -B package -Ddir=/tmp/codecrafters-shell-target && sed -i 's/^\(mvn .*\)/#\1/' ./your_program.sh" > /codecrafters-precompile.sh
```
Comment (rohitpaulk) [inline dockerfiles/java-21.Dockerfile:18]: @glokta1 this will affect existing repositories. Another example of not thinking about compatibility

**Review by rohitpaulk — CHANGES_REQUESTED** [review]
Added one round of comments

File: compiled_starters/java/src/main/java/Main.java (hunk 6 above)
Comment (glokta1) [reply compiled_starters/java/src/main/java/Main.java:8]: @rohitpaulk Yes. It wasn't consistent with other challenges – to pass first stage, they only need to uncomment the line that prints the `$ ` prompt.

File: compiled_starters/c/.codecrafters/compile.sh [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+#
+# This script is used to compile your program on CodeCrafters
+# 
+# This runs before .codecrafters/run.sh
+#
+# Learn more: codecrafters.io/program-interface
```
Comment (rohitpaulk) [inline compiled_starters/c/.codecrafters/compile.sh:7]:
```suggestion
# Learn more: https://codecrafters.io/program-interface
```

File: compiled_starters/c/app/main.c [hunk 12]
```diff
@@ -1,6 +1,7 @@
 #include <stdio.h>
 
-int main() {
+int main()
```
Comment (rohitpaulk) [inline compiled_starters/c/app/main.c:3]: Unrelated change

File: compiled_starters/c/your_program.sh [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+# Learn more: codecrafters.io/program-interface
```
Comment (rohitpaulk) [inline compiled_starters/c/your_program.sh:8]: Same as above, add link

File: compiled_starters/cpp/.codecrafters/compile.sh [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+# 
+# This runs before .codecrafters/run.sh
+#
+# Learn more: codecrafters.io/program-interface
+
+# Exit early if any commands fail
+set -e
+
+cmake -B build -S . -DCMAKE_TOOLCHAIN_FILE=${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake
+cmake --build ./build
```
Comment (rohitpaulk) [inline compiled_starters/cpp/.codecrafters/compile.sh:13]: Needs `exec` at the end - else signals sent to the shell script won't be forwarded to the child program

File: dockerfiles/cpp-23.Dockerfile [hunk 15]
```diff
@@ -38,8 +38,10 @@ RUN mkdir -p /app-cached/build
 RUN if [ -d "/app/build" ]; then mv /app/build /app-cached; fi
 RUN if [ -d "/app/vcpkg_installed" ]; then mv /app/vcpkg_installed /app-cached/build; fi
 
-RUN echo "cd \${CODECRAFTERS_SUBMISSION_DIR} && cmake -B build -S . -DCMAKE_TOOLCHAIN_FILE=${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake && cmake --build ./build && sed -i '/^cmake/ s/^/# /' ./your_shell.sh" > /codecrafters-precompile.sh
-RUN chmod +x /codecrafters-precompile.sh
+RUN if [ -f "$CODECRAFTERS_SUBMISSION_DIR/your_shell.sh" ]; then \
+    echo "cd \${CODECRAFTERS_SUBMISSION_DIR} && cmake -B build -S . -DCMAKE_TOOLCHAIN_FILE=${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake && cmake --build ./build && sed -i '/^cmake/ s/^/# /' ./your_program.sh" > /codecrafters-precompile.sh && \
```
Comment (rohitpaulk) [inline dockerfiles/cpp-23.Dockerfile:42]: Can keep this as-is, won't run unless they delete run.sh
